### ISSUE TITLE
Heap corrupted by regex involving (?| ... ) and ($0)

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#12096]
Migrated from [rt.perl.org#112874](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=112874) (status was 'resolved')

Searchable as RT112874$

**Comment by p5pRT** [comment]
### From @jimav


This is a bug report for perl from james\_avera@&#8203;yahoo\.com\,
generated with the help of perlbug 1\.39 running under perl 5\.12\.4\.



---
The following Perl program causes a fatal error in glibc indicating
that the heap is probably corrupted\.  The error messages are written
directly to the terminal \(probably /dev/tty\)\, and does not appear on
either stdout or stderr\.

\#\!/usr/bin/perl
use strict; use warnings;

$\_ = 'aaa x=$\{ \\$x \} bbb a=@&#8203;a=@&#8203;\{ \\@&#8203;a \} ccc';
s\{
&nbsp;   \(\[\\$\\@&#8203;\]\)
&nbsp;   \(?|   \# re\-use $1\, $2 etc in each alternation
&nbsp;     \(\[^\\w\\s\\\{\]\)
&nbsp;     | \( \\\{
&nbsp;            \(
&nbsp;              \(?> \\\\\.
&nbsp;                  | \[^\\\{\\\}\\\\\]\+\+
&nbsp;                  | \\\{ \(?0\) \\\}
&nbsp;              \)\*
&nbsp;            \)
&nbsp;         \\\}
&nbsp;       \)
&nbsp;   \)
&nbsp;  \}
&nbsp;  \{replacement\}xgs;


Here is what appears on the /dev/tty \(captured with 'script'\)&#8203;:

&nbsp;   \*\*\* glibc detected \*\*\* /usr/bin/perl&#8203;: free\(\)&#8203;: invalid next size 
\(fast\)&#8203;: 0x0000000001198890 \*\*\*
&nbsp;   ======= Backtrace&#8203;: =========
&nbsp;   /lib/x86\_64\-linux\-gnu/libc\.so\.6\(\+0x7a6e6\)\[0x7f6ca7fdf6e6\]
&nbsp;   /lib/x86\_64\-linux\-gnu/libc\.so\.6\(cfree\+0x6c\)\[0x7f6ca7fe39cc\]
&nbsp;   /usr/lib/libperl\.so\.5\.12\(Perl\_regexec\_flags\+0x41c\)\[0x7f6ca8419e1c\]
&nbsp;   /usr/lib/libperl\.so\.5\.12\(Perl\_pp\_subst\+0x939\)\[0x7f6ca83b42c9\]
&nbsp;   /usr/lib/libperl\.so\.5\.12\(Perl\_runops\_standard\+0x20\)\[0x7f6ca83adc10\]
&nbsp;   /usr/lib/libperl\.so\.5\.12\(perl\_run\+0x35e\)\[0x7f6ca835782e\]
&nbsp;   /usr/bin/perl\(main\+0x149\)\[0x400d49\]
&nbsp;   /lib/x86\_64\-linux\-gnu/libc\.so\.6\(\_\_libc\_start\_main\+0xed\)\[0x7f6ca7f8630d\]
&nbsp;   /usr/bin/perl\[0x400d81\]
&nbsp;   ======= Memory map&#8203;: ========
&nbsp;   00400000\-00402000 r\-xp 00000000 08&#8203;:15 1851 
&nbsp;     /usr/bin/perl
&nbsp;   00601000\-00602000 r\-\-p 00001000 08&#8203;:15 1851 
&nbsp;     /usr/bin/perl
&nbsp;   00602000\-00603000 rw\-p 00002000 08&#8203;:15 1851 
&nbsp;     /usr/bin/perl
&nbsp;   0116c000\-011cf000 rw\-p 00000000 00&#8203;:00 0 
&nbsp;     \[heap\]
&nbsp;   7f6ca0000000\-7f6ca0021000 rw\-p 00000000 00&#8203;:00 0
&nbsp;   7f6ca0021000\-7f6ca4000000 \-\-\-p 00000000 00&#8203;:00 0
&nbsp;   7f6ca6d88000\-7f6ca6d9d000 r\-xp 00000000 08&#8203;:15 137088 
&nbsp;     /lib/x86\_64\-linux\-gnu/libgcc\_s\.so\.1
&nbsp;   7f6ca6d9d000\-7f6ca6f9c000 \-\-\-p 00015000 08&#8203;:15 137088 
&nbsp;     /lib/x86\_64\-linux\-gnu/libgcc\_s\.so\.1
&nbsp;   7f6ca6f9c000\-7f6ca6f9d000 r\-\-p 00014000 08&#8203;:15 137088 
&nbsp;     /lib/x86\_64\-linux\-gnu/libgcc\_s\.so\.1
&nbsp;   7f6ca6f9d000\-7f6ca6f9e000 rw\-p 00015000 08&#8203;:15 137088 
&nbsp;     /lib/x86\_64\-linux\-gnu/libgcc\_s\.so\.1
&nbsp;   7f6ca6f9e000\-7f6ca7687000 r\-\-p 00000000 08&#8203;:15 7656 
&nbsp;     /usr/lib/locale/locale\-archive
&nbsp;   7f6ca7687000\-7f6ca7690000 r\-xp 00000000 08&#8203;:15 131553 
&nbsp;     /lib/x86\_64\-linux\-gnu/libcrypt\-2\.13\.so
&nbsp;   7f6ca7690000\-7f6ca7890000 \-\-\-p 00009000 08&#8203;:15 131553 
&nbsp;     /lib/x86\_64\-linux\-gnu/libcrypt\-2\.13\.so
&nbsp;   7f6ca7890000\-7f6ca7891000 r\-\-p 00009000 08&#8203;:15 131553 
&nbsp;     /lib/x86\_64\-linux\-gnu/libcrypt\-2\.13\.so
&nbsp;   7f6ca7891000\-7f6ca7892000 rw\-p 0000a000 08&#8203;:15 131553 
&nbsp;     /lib/x86\_64\-linux\-gnu/libcrypt\-2\.13\.so
&nbsp;   7f6ca7892000\-7f6ca78c0000 rw\-p 00000000 00&#8203;:00 0
&nbsp;   7f6ca78c0000\-7f6ca78d8000 r\-xp 00000000 08&#8203;:15 131576 
&nbsp;     /lib/x86\_64\-linux\-gnu/libpthread\-2\.13\.so
&nbsp;   7f6ca78d8000\-7f6ca7ad7000 \-\-\-p 00018000 08&#8203;:15 131576 
&nbsp;     /lib/x86\_64\-linux\-gnu/libpthread\-2\.13\.so
&nbsp;   7f6ca7ad7000\-7f6ca7ad8000 r\-\-p 00017000 08&#8203;:15 131576 
&nbsp;     /lib/x86\_64\-linux\-gnu/libpthread\-2\.13\.so
&nbsp;   7f6ca7ad8000\-7f6ca7ad9000 rw\-p 00018000 08&#8203;:15 131576 
&nbsp;     /lib/x86\_64\-linux\-gnu/libpthread\-2\.13\.so
&nbsp;   7f6ca7ad9000\-7f6ca7add000 rw\-p 00000000 00&#8203;:00 0
&nbsp;   7f6ca7add000\-7f6ca7b60000 r\-xp 00000000 08&#8203;:15 131557 
&nbsp;     /lib/x86\_64\-linux\-gnu/libm\-2\.13\.so
&nbsp;   7f6ca7b60000\-7f6ca7d5f000 \-\-\-p 00083000 08&#8203;:15 131557 
&nbsp;     /lib/x86\_64\-linux\-gnu/libm\-2\.13\.so
&nbsp;   7f6ca7d5f000\-7f6ca7d60000 r\-\-p 00082000 08&#8203;:15 131557 
&nbsp;     /lib/x86\_64\-linux\-gnu/libm\-2\.13\.so
&nbsp;   7f6ca7d60000\-7f6ca7d61000 rw\-p 00083000 08&#8203;:15 131557 
&nbsp;     /lib/x86\_64\-linux\-gnu/libm\-2\.13\.so
&nbsp;   7f6ca7d61000\-7f6ca7d63000 r\-xp 00000000 08&#8203;:15 131555 
&nbsp;     /lib/x86\_64\-linux\-gnu/libdl\-2\.13\.so
&nbsp;   7f6ca7d63000\-7f6ca7f63000 \-\-\-p 00002000 08&#8203;:15 131555 
&nbsp;     /lib/x86\_64\-linux\-gnu/libdl\-2\.13\.so
&nbsp;   7f6ca7f63000\-7f6ca7f64000 r\-\-p 00002000 08&#8203;:15 131555 
&nbsp;     /lib/x86\_64\-linux\-gnu/libdl\-2\.13\.so
&nbsp;   7f6ca7f64000\-7f6ca7f65000 rw\-p 00003000 08&#8203;:15 131555 
&nbsp;     /lib/x86\_64\-linux\-gnu/libdl\-2\.13\.so
&nbsp;   7f6ca7f65000\-7f6ca80fc000 r\-xp 00000000 08&#8203;:15 131550 
&nbsp;     /lib/x86\_64\-linux\-gnu/libc\-2\.13\.so
&nbsp;   7f6ca80fc000\-7f6ca82fb000 \-\-\-p 00197000 08&#8203;:15 131550 
&nbsp;     /lib/x86\_64\-linux\-gnu/libc\-2\.13\.so
&nbsp;   7f6ca82fb000\-7f6ca82ff000 r\-\-p 00196000 08&#8203;:15 131550 
&nbsp;     /lib/x86\_64\-linux\-gnu/libc\-2\.13\.so
&nbsp;   7f6ca82ff000\-7f6ca8300000 rw\-p 0019a000 08&#8203;:15 131550 
&nbsp;     /lib/x86\_64\-linux\-gnu/libc\-2\.13\.so
&nbsp;   7f6ca8300000\-7f6ca8306000 rw\-p 00000000 00&#8203;:00 0
&nbsp;   7f6ca8306000\-7f6ca8472000 r\-xp 00000000 08&#8203;:15 5423 
&nbsp;     /usr/lib/libperl\.so\.5\.12\.4
&nbsp;   7f6ca8472000\-7f6ca8671000 \-\-\-p 0016c000 08&#8203;:15 5423 
&nbsp;     /usr/lib/libperl\.so\.5\.12\.4
&nbsp;   7f6ca8671000\-7f6ca8675000 r\-\-p 0016b000 08&#8203;:15 5423 
&nbsp;     /usr/lib/libperl\.so\.5\.12\.4
&nbsp;   7f6ca8675000\-7f6ca867a000 rw\-p 0016f000 08&#8203;:15 5423 
&nbsp;     /usr/lib/libperl\.so\.5\.12\.4
&nbsp;   7f6ca867a000\-7f6ca869b000 r\-xp 00000000 08&#8203;:15 131226 
&nbsp;     /lib/x86\_64\-linux\-gnu/ld\-2\.13\.so
&nbsp;   7f6ca8871000\-7f6ca8876000 rw\-p 00000000 00&#8203;:00 0
&nbsp;   7f6ca8898000\-7f6ca889a000 rw\-p 00000000 00&#8203;:00 0
&nbsp;   7f6ca889a000\-7f6ca889b000 r\-\-p 00020000 08&#8203;:15 131226 
&nbsp;     /lib/x86\_64\-linux\-gnu/ld\-2\.13\.so
&nbsp;   7f6ca889b000\-7f6ca889d000 rw\-p 00021000 08&#8203;:15 131226 
&nbsp;     /lib/x86\_64\-linux\-gnu/ld\-2\.13\.so
&nbsp;   7fffc1db5000\-7fffc1dd6000 rw\-p 00000000 00&#8203;:00 0 
&nbsp;     \[stack\]
&nbsp;   7fffc1dff000\-7fffc1e00000 r\-xp 00000000 00&#8203;:00 0 
&nbsp;     \[vdso\]
&nbsp;   ffffffffff600000\-ffffffffff601000 r\-xp 00000000 00&#8203;:00 0 
&nbsp;     \[vsyscall\]
&nbsp;   Aborted



---


---
Flags&#8203;:
&nbsp;     category=core
&nbsp;     severity=high


---
Site configuration information for perl 5\.12\.4&#8203;:

Configured by Debian Project at Tue Sep  6 08&#8203;:08&#8203;:24 UTC 2011\.

Summary of my perl5 \(revision 5 version 12 subversion 4\) configuration&#8203;:

&nbsp;   Platform&#8203;:
&nbsp;     osname=linux\, osvers=2\.6\.24\-28\-server\, 
archname=x86\_64\-linux\-gnu\-thread\-multi
&nbsp;     uname='linux allspice 2\.6\.24\-28\-server \#1 smp wed aug 18 21&#8203;:17&#8203;:51 
utc 2010 x86\_64 x86\_64 x86\_64 gnulinux '
&nbsp;     config\_args='\-Dusethreads \-Duselargefiles \-Dccflags=\-DDEBIAN 
\-Dcccdlflags=\-fPIC \-Darchname=x86\_64\-linux\-gnu \-Dprefix=/usr 
\-Dprivlib=/usr/share/perl/5\.12 \-Darchlib=/usr/lib/perl/5\.12 
\-Dvendorprefix=/usr \-Dvendorlib=/usr/share/perl5 
\-Dvendorarch=/usr/lib/perl5 \-Dsiteprefix=/usr/local 
\-Dsitelib=/usr/local/share/perl/5\.12\.4 
\-Dsitearch=/usr/local/lib/perl/5\.12\.4 \-Dman1dir=/usr/share/man/man1 
\-Dman3dir=/usr/share/man/man3 \-Dsiteman1dir=/usr/local/man/man1 
\-Dsiteman3dir=/usr/local/man/man3 \-Duse64bitint \-Dman1ext=1 
\-Dman3ext=3perl \-Dpager=/usr/bin/sensible\-pager \-Uafs \-Ud\_csh \-Ud\_ualarm 
\-Uusesfio \-Uusenm \-Ui\_libutil \-DDEBUGGING=\-g \-Doptimize=\-O2 \-Duseshrplib 
\-Dlibperl=libperl\.so\.5\.12\.4 \-des'
&nbsp;     hint=recommended\, useposix=true\, d\_sigaction=define
&nbsp;     useithreads=define\, usemultiplicity=define
&nbsp;     useperlio=define\, d\_sfio=undef\, uselargefiles=define\, usesocks=undef
&nbsp;     use64bitint=define\, use64bitall=define\, uselongdouble=undef
&nbsp;     usemymalloc=n\, bincompat5005=undef
&nbsp;   Compiler&#8203;:
&nbsp;     cc='cc'\, ccflags ='\-D\_REENTRANT \-D\_GNU\_SOURCE \-DDEBIAN 
\-fno\-strict\-aliasing \-pipe \-fstack\-protector \-I/usr/local/include 
\-D\_LARGEFILE\_SOURCE \-D\_FILE\_OFFSET\_BITS=64'\,
&nbsp;     optimize='\-O2 \-g'\,
&nbsp;     cppflags='\-D\_REENTRANT \-D\_GNU\_SOURCE \-DDEBIAN \-fno\-strict\-aliasing 
\-pipe \-fstack\-protector \-I/usr/local/include'
&nbsp;     ccversion=''\, gccversion='4\.6\.1'\, gccosandvers=''
&nbsp;     intsize=4\, longsize=8\, ptrsize=8\, doublesize=8\, byteorder=12345678
&nbsp;     d\_longlong=define\, longlongsize=8\, d\_longdbl=define\, longdblsize=16
&nbsp;     ivtype='long'\, ivsize=8\, nvtype='double'\, nvsize=8\, Off\_t='off\_t'\, 
lseeksize=8
&nbsp;     alignbytes=8\, prototype=define
&nbsp;   Linker and Libraries&#8203;:
&nbsp;     ld='cc'\, ldflags =' \-fstack\-protector \-L/usr/local/lib'
&nbsp;     libpth=/usr/local/lib /lib/x86\_64\-linux\-gnu /lib/\.\./lib 
/usr/lib/x86\_64\-linux\-gnu /usr/lib/\.\./lib /lib /usr/lib
&nbsp;     libs=\-lgdbm \-lgdbm\_compat \-ldb \-ldl \-lm \-lpthread \-lc \-lcrypt
&nbsp;     perllibs=\-ldl \-lm \-lpthread \-lc \-lcrypt
&nbsp;     libc=\, so=so\, useshrplib=true\, libperl=libperl\.so\.5\.12\.4
&nbsp;     gnulibc\_version='2\.13'
&nbsp;   Dynamic Linking&#8203;:
&nbsp;     dlsrc=dl\_dlopen\.xs\, dlext=so\, d\_dlsymun=undef\, ccdlflags='\-Wl\,\-E'
&nbsp;     cccdlflags='\-fPIC'\, lddlflags='\-shared \-O2 \-g \-L/usr/local/lib 
\-fstack\-protector'

Locally applied patches&#8203;:




---
@&#8203;INC for perl 5\.12\.4&#8203;:
&nbsp;     /home/jima/local/share/perl/5\.12\.4
&nbsp;     /home/jima/local/share/perl
&nbsp;     /home/jima/lib/perl
&nbsp;     /etc/perl
&nbsp;     /usr/local/lib/perl/5\.12\.4
&nbsp;     /usr/local/share/perl/5\.12\.4
&nbsp;     /usr/lib/perl5
&nbsp;     /usr/share/perl5
&nbsp;     /usr/lib/perl/5\.12
&nbsp;     /usr/share/perl/5\.12
&nbsp;     /usr/local/lib/site\_perl
&nbsp;     \.



---
Environment for perl 5\.12\.4&#8203;:
&nbsp;     HOME=/home/jima
&nbsp;     LANG=en\_US\.UTF\-8
&nbsp;     LANGUAGE \(unset\)
&nbsp;     LD\_LIBRARY\_PATH=/home/jima/local/lib
&nbsp;     LOGDIR \(unset\)
 
PATH=/home/jima/bin&#8203;:/home/jima/local/bin&#8203;:/home/jima/jima\_tools/linux86\_64/bin&#8203;:/usr/bin&#8203;:/bin&#8203;:/usr/sbin&#8203;:/sbin&#8203;:/usr/bin/X11&#8203;:/usr/local/bin&#8203;:/opt/openoffice\.org3/program&#8203;:/usr/lib/lightdm/lightdm&#8203;:/usr/local/sbin&#8203;:/usr/games&#8203;:\.
&nbsp;     PERL5LIB=/home/jima/local/share/perl&#8203;:/home/jima/lib/perl
&nbsp;     PERL\_BADLANG \(unset\)
&nbsp;     SHELL=/bin/bash

**Comment by p5pRT** [comment]
### From @jimav

Sending updated bug demo\, as an attachment this time\.

If the alternations inside \(?| \.\.\. \) are swapped\, or a 3rd alternation
added\, the crash does not happen\.  See comment in the attached

**Comment by p5pRT** [comment]
### From @jimav

[bugdemo.pl](https://rt-archive.perl.org/perl5/Ticket/Attachment/1115004/561770/bugdemo.pl)

**Comment by p5pRT** [comment]
### From [Unknown Contact. See original ticket]

Sending updated bug demo\, as an attachment this time\.

If the alternations inside \(?| \.\.\. \) are swapped\, or a 3rd alternation
added\, the crash does not happen\.  See comment in the attached

**Comment by p5pRT** [comment]
@jimav - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From @iabyn

On Fri\, May 11\, 2012 at 04&#8203;:02&#8203;:15PM \-0700\, Jim Avera wrote&#8203;:
> The following Perl program causes a fatal error in glibc indicating
> that the heap is probably corrupted\.  The error messages are written
> directly to the terminal \(probably /dev/tty\)\, and does not appear on
> either stdout or stderr\.
> 
> \#\!/usr/bin/perl
> use strict; use warnings;
> 
> $\_ = 'aaa x=$\{ \\$x \} bbb a=@&#8203;a=@&#8203;\{ \\@&#8203;a \} ccc';
> s\{
>    \(\[\\$\\@&#8203;\]\)
>    \(?|   \# re\-use $1\, $2 etc in each alternation
>      \(\[^\\w\\s\\\{\]\)
>      | \( \\\{
>             \(
>               \(?> \\\\\.
>                   | \[^\\\{\\\}\\\\\]\+\+
>                   | \\\{ \(?0\) \\\}
>               \)\*
>             \)
>          \\\}
>        \)
>    \)
>   \}
>   \{replacement\}xgs;

valgrind shows it crashing on 5\.12\.4 but not on 5\.14\.0 or later\.

\-\- 
Red sky at night \- gerroff my land\!
Red sky at morning \- gerroff my land\!
&nbsp;    \-\- old farmers' sayings \#14

**Comment by p5pRT** [comment]
### From @demerphq

On 13 May 2012 23&#8203;:40\, Dave Mitchell \<davem@&#8203;iabyn\.com> wrote&#8203;:
> On Fri\, May 11\, 2012 at 04&#8203;:02&#8203;:15PM \-0700\, Jim Avera wrote&#8203;:
>> The following Perl program causes a fatal error in glibc indicating
>> that the heap is probably corrupted\.  The error messages are written
>> directly to the terminal \(probably /dev/tty\)\, and does not appear on
>> either stdout or stderr\.
>>
>> \#\!/usr/bin/perl
>> use strict; use warnings;
>>
>> $\_ = 'aaa x=$\{ \\$x \} bbb a=@&#8203;a=@&#8203;\{ \\@&#8203;a \} ccc';
>> s\{
>>    \(\[\\$\\@&#8203;\]\)
>>    \(?|   \# re\-use $1\, $2 etc in each alternation
>>      \(\[^\\w\\s\\\{\]\)
>>      | \( \\\{
>>             \(
>>               \(?> \\\\\.
>>                   | \[^\\\{\\\}\\\\\]\+\+
>>                   | \\\{ \(?0\) \\\}
>>               \)\*
>>             \)
>>          \\\}
>>        \)
>>    \)
>>   \}
>>   \{replacement\}xgs;
>
> valgrind shows it crashing on 5\.12\.4 but not on 5\.14\.0 or later\.

I seem to recall a fix being applied\, but I do not remember the
details of who wrote or applied it\. Maybe FC?

I think it had to do with miscounting the number of capture buffers\.

cheers\,
Yves

\-\- 
perl \-Mre=debug \-e "/just|another|perl|hacker/"

**Comment by p5pRT** [comment]
### From @nwc10

On Mon\, May 14\, 2012 at 11&#8203;:18&#8203;:18AM \+0200\, demerphq wrote&#8203;:
> On 13 May 2012 23&#8203;:40\, Dave Mitchell \<davem@&#8203;iabyn\.com> wrote&#8203;:
> > On Fri\, May 11\, 2012 at 04&#8203;:02&#8203;:15PM \-0700\, Jim Avera wrote&#8203;:
> >> The following Perl program causes a fatal error in glibc indicating
> >> that the heap is probably corrupted\.  The error messages are written
> >> directly to the terminal \(probably /dev/tty\)\, and does not appear on
> >> either stdout or stderr\.
> >>
> >> \#\!/usr/bin/perl
> >> use strict; use warnings;
> >>
> >> $\_ = 'aaa x=$\{ \\$x \} bbb a=@&#8203;a=@&#8203;\{ \\@&#8203;a \} ccc';
> >> s\{
> >>    \(\[\\$\\@&#8203;\]\)
> >>    \(?|   \# re\-use $1\, $2 etc in each alternation
> >>      \(\[^\\w\\s\\\{\]\)
> >>      | \( \\\{
> >>             \(
> >>               \(?> \\\\\.
> >>                   | \[^\\\{\\\}\\\\\]\+\+
> >>                   | \\\{ \(?0\) \\\}
> >>               \)\*
> >>             \)
> >>          \\\}
> >>        \)
> >>    \)
> >>   \}
> >>   \{replacement\}xgs;
> >
> > valgrind shows it crashing on 5\.12\.4 but not on 5\.14\.0 or later\.
> 
> I seem to recall a fix being applied\, but I do not remember the
> details of who wrote or applied it\. Maybe FC?
> 
> I think it had to do with miscounting the number of capture buffers\.

Fixed by&#8203;:

commit fd4be6f07df0e6a021290ef721c5d73550e0248c
Author&#8203;: Father Chrysostomos \<sprout@&#8203;cpan\.org>
Date&#8203;:   Fri Feb 25 20&#8203;:45&#8203;:08 2011 \-0800

&nbsp;    \[perl \#84746\] Accessing $2 causes the interpreter to crash
&nbsp;    
&nbsp;    Actually\, it doesn't\. The original test case was&#8203;:
&nbsp;    
&nbsp;    \#\!/usr/bin/perl
&nbsp;    
&nbsp;    my $rx = qr'\\$ \(?| \{\(\.\+?\)\} | \(\.\+?\); | \(\.\+?\)\(\\s\) \)'x;
&nbsp;    my $test = '/home/$USERNAME ';
&nbsp;    die unless $test =~ $rx;
&nbsp;    print "1&#8203;: $1\\n";
&nbsp;    print "2&#8203;: $2\\n" if defined $2;
&nbsp;    
&nbsp;    This crashes even if I put an 'exit' right after the pattern match\.
&nbsp;    
&nbsp;    What's happening is that regcomp miscounts the number of capturing
&nbsp;    parenthesis pairs \(cf\. \[perl \#59734\]\)\, so the execution of the regular
&nbsp;    expression causes a buffer overflow which overwrites the op\_sibling
&nbsp;    field of the regcreset op\, causing a crash when the op is freed\. \(The
&nbsp;    exact failure may differ between builds\, platforms\, etc\.\, of course\.\)
&nbsp;    
&nbsp;    S\_reg in regcomp\.c keeps a count of the parenthesised groups in a
&nbsp;    \(?|\.\.\.\) construct\, which it updates after each branch\, if that branch
&nbsp;    has more captures than any previous branch\. But it was not updating
&nbsp;    the count after the last branch\.
&nbsp;    
&nbsp;    So this bug would occur if the last branch had more capturing paren\-
&nbsp;    theses than any previous branch\.
&nbsp;    
&nbsp;    Commit ee91d26\, which fixed bug \#59734\, only solved the problem when
&nbsp;    there was just one branch \(by updating the count before the loop that
&nbsp;    deals with subsequent branches was entered\)\.
&nbsp;    
&nbsp;    This commit changes the code at the end of S\_reg to take into account
&nbsp;    that RExC\_npar \(the current paren count\) might have been increased by
&nbsp;    the last branch\.
&nbsp;    
&nbsp;    Since the loop to deal with subsequent branches resets the count
&nbsp;    \*before\* each branch\, the code that commit ee91d26 added is no longer
&nbsp;    necessary\, so this commit removes it\.

<details><summary>Inline Patch</summary>

```diff
diff --git a/regcomp.c b/regcomp.c
index 96a825c..932da1a 100644
--- a/regcomp.c
+++ b/regcomp.c
@@ -7097,12 +7097,6 @@ S_reg(pTHX_ RExC_state_t *pRExC_state, I32 paren, I32 *flagp,U32 depth)
     parse_start = RExC_parse;   /* MJD */
     br = regbranch(pRExC_state, &flags, 1,depth+1);
 
-    if (freeze_paren) {
-        if (RExC_npar > after_freeze)
-            after_freeze = RExC_npar;
-        RExC_npar = freeze_paren;
-    }
-
     /*     branch_len = (paren != 0); */
 
     if (br == NULL)
@@ -7246,7 +7240,7 @@ S_reg(pTHX_ RExC_state_t *pRExC_state, I32 paren, I32 *flagp,U32 depth)
     if (RExC_in_lookbehind) {
 	RExC_in_lookbehind--;
     }
-    if (after_freeze)
+    if (after_freeze > RExC_npar)
         RExC_npar = after_freeze;
     return(ret);
 }
diff --git a/t/re/re_tests b/t/re/re_tests
index 5441437..924434c 100644
--- a/t/re/re_tests
+++ b/t/re/re_tests
@@ -1329,6 +1329,9 @@ X(\w+)(?=\s)|X(\w+)	Xab	y	[$1-$2]	[-ab]
 (?|(?<foo>x)|(?<bar>y))	x	yM	$+{foo}	x	miniperl cannot load Tie::Hash::NamedCapture
 (?|(?<bar>y)|(?<foo>x))	x	yM	$+{foo}	x	miniperl cannot load Tie::Hash::NamedCapture
 (?<bar>)(?|(?<foo>x))	x	yM	$+{foo}	x	miniperl cannot load Tie::Hash::NamedCapture
+# Used to crash, because the last branch was ignored when the parens
+# were counted:
+(?|(b)|()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()(a))	a	y	$&	a
 
 #Bug #41492
 (?(DEFINE)(?<A>(?&B)+)(?<B>a))(?&A)	a	y	$&	a



```
</details>
git bisect is everyone's friend.

Nicholas Clark

**Comment by p5pRT** [comment]
@iabyn - Status changed from 'open' to 'resolved'
